### PR TITLE
Storages: disaggregated mode supports storage read thread.

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: ref #6834

### What is changed and how it works?
- Refine `RNWorkerFetchPages` and `RNWorkerPrepareStreams`，encapsulate the functions into `SegmentReadTask` for reuse.
- The read thread pool supports disaggregated mode.
- Add some unittests.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/Remote/RNWorkerFetchPages.h [hunk 1]
```diff
@@ -31,13 +31,14 @@ class RNWorkerFetchPages
     , public ThreadedWorker<SegmentReadTaskPtr, SegmentReadTaskPtr>
 {
 protected:
-    SegmentReadTaskPtr doWork(const SegmentReadTaskPtr & task) override;
+    SegmentReadTaskPtr doWork(const SegmentReadTaskPtr & task) override
+    {
+        task->fetchPages();
+        return task;
+    }
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/Remote/RNWorkerFetchPages.h:38]: seems not used?

File: dbms/src/Storages/DeltaMerge/Remote/RNWorkerPrepareStreams.h [hunk 2]
```diff
@@ -35,7 +35,18 @@ class RNWorkerPrepareStreams
...
+        const auto & settings = task->dm_context->db_context.getSettingsRef();
+        task->initInputStream(
+            *columns_to_read,
+            read_tso,
+            push_down_filter,
+            read_mode,
+            settings.max_block_size,
+            settings.dt_enable_delta_index_error_fallback);
+        return task;
+    }
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/Remote/RNWorkerPrepareStreams.h:49]: seems not used?

File: dbms/src/Storages/DeltaMerge/Remote/Serializer.cpp [hunk 3]
```diff
@@ -112,6 +115,41 @@ RemotePb::RemoteSegment Serializer::serializeTo(
...
+}
+#endif
+
+static DMFilePtr restoreDMFile(const Remote::IDataStorePtr & data_store, const String & remote_fname)
+{
+#ifndef DBMS_PUBLIC_GTEST
+    auto prepared = data_store->prepareDMFileByKey(remote_fname);
+    auto dmfile = prepared->restore(DMFile::ReadMetaMode::all());
+#else
+    (void)data_store; // Disable warnning of unused parameter.
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/Remote/Serializer.cpp:139]: mark `[[maybe_unused]]` instead?

File: dbms/src/Storages/DeltaMerge/tests/gtest_dm_delta_merge_store_test_basic.h [hunk 4]
```diff
@@ -89,6 +89,31 @@ class DeltaMergeStoreTest : public DB::base::TiFlashStorageTestBasic
...
+
+        const auto & pk_column = block.getByPosition(0).column;
+        auto min_pk = pk_column->getInt(0);
+        auto max_pk = pk_column->getInt(block.rows() - 1);
+        HandleRange range(min_pk, max_pk + 1);
+        auto handle_range = RowKeyRange::fromHandleRange(range);
+        auto external_file = ExternalDTFileInfo{.id = file_id, .range = handle_range};
+        return {
+            handle_range,
+            {external_file}}; // There are some duplicated info. This is to minimize the change to our test code.
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/tests/gtest_dm_delta_merge_store_test_basic.h:114]:
```suggestion
        // There are some duplicated info. This is to minimize the change to our test code.
        return {handle_range, {external_file}}; 
```

File: dbms/src/Storages/DeltaMerge/Remote/RNWorkerFetchPages.h (hunk 1 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Remote/RNWorkerFetchPages.h:38]: get

File: dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp [hunk 5]
```diff
@@ -229,7 +302,298 @@ void SegmentReadTask::initInputStream(
...
+// and send them to `RNWritePageCachePool` to write into local page cache.
+struct WritePageTask
+{
+    explicit WritePageTask(Remote::RNLocalPageCache * page_cache_)
+        : page_cache(page_cache_)
+    {}
+    Remote::RNLocalPageCache * page_cache;
+    UniversalWriteBatch wb;
+    std::list<DM::RemotePb::RemotePage> remote_pages; // Hold the data of wb.
+    std::list<MemTrackerWrapper> remote_page_mem_tracker_wrappers; // Hold the memory stat of remote_pages.
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp:443]: why use `std::list` here?

File: dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp [hunk 6]
```diff
@@ -229,7 +302,298 @@ void SegmentReadTask::initInputStream(
...
+    WritePageTaskPtr write_page_task;
+    std::vector<std::future<void>> write_page_results;
+
+    // Keep reading packets.
+    while (true)
+    {
+        Stopwatch sw_read_packet;
+        auto packet = std::make_shared<disaggregated::PagesPacket>();
+        if (bool more = stream_resp->Read(packet.get()); !more)
+            break;
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp:498]:
```suggestion
        if (!stream_resp->Read(packet.get()))
            break;
```

File: dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp [hunk 7]
```diff
@@ -229,7 +302,298 @@ void SegmentReadTask::initInputStream(
...
+            };
+            auto read_buffer
+                = std::make_shared<ReadBufferFromMemory>(remote_page.data().data(), remote_page.data().size());
+            PageFieldSizes field_sizes;
+            field_sizes.reserve(remote_page.field_sizes_size());
+            for (const auto & field_sz : remote_page.field_sizes())
+            {
+                field_sizes.emplace_back(field_sz);
+            }
+            deserialize_page_ns += sw.elapsed();
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp:545]: just use `elapsedMilliseconds()`

File: dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp (hunk 5 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp:443]: how about `std::forward_list`?

File: dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp (hunk 7 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp:545]: get

**Review by Lloyd-Pottiger — APPROVED** [review]

File: dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.cpp [hunk 8]
```diff
@@ -47,6 +47,7 @@ void MergedTask::initOnce()
             setStreamFinished(cur_idx);
             continue;
         }
+        task->fetchPages();
         stream = pool->buildInputStream(task);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.cpp:51]: Should we move `task->fetchPages` into `pool->buildInputStream`?

File: dbms/src/Storages/DeltaMerge/Remote/Serializer.cpp [hunk 9]
```diff
@@ -112,6 +115,40 @@ RemotePb::RemoteSegment Serializer::serializeTo(
...
+    auto file_provider = std::make_shared<FileProvider>(key_manager, false);
+    auto dmfile = DMFile::restore(
+        file_provider,
+        file_id,
+        /*page_id*/ 0,
+        parent_path,
+        DMFile::ReadMetaMode::all());
+#endif
+    return dmfile;
+}
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Remote/Serializer.cpp:150]: I think this is too hacky. Maybe write a `class MockDataStore` and handle the testing logic there.
Then we can make the data_store as one param of `Serializer::deserializeSegmentSnapshotFrom` to write the testing logic.

File: dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp [hunk 10]
```diff
@@ -214,13 +223,77 @@ void SegmentReadTask::initColumnFileDataProvider(const Remote::RNLocalPageCacheG
...
+}
+
+bool SegmentReadTask::doInitInputStream(
+    const ColumnDefines & columns_to_read,
+    UInt64 read_tso,
+    const PushDownFilterPtr & push_down_filter,
+    ReadMode read_mode,
+    size_t expected_block_size,
+    bool enable_delta_index_error_fallback)
+{
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp:263]: This function has the same name as line 283, which is confusing.
Better rename it to `SegmentReadTask::tryInitInputStreamWithFallback(...`

**Comment by JaySon-Huang** [comment]
Suggest some changes here: https://github.com/JinheLin/tiflash-1/pull/1

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**